src/test/groovy/graphql/validation/rules/DeferDirectiveOnValidOperationTest.groovy

```diff
@@ -0,0 +1,252 @@
+    ValidationErrorCollector errorCollector = new ValidationErrorCollector()
+
+    def traverse(String query) {
+        Document document = new Parser().parseDocument(query)
+        I18n i18n = I18n.i18n(I18n.BundleType.Validation, Locale.ENGLISH)
+        ValidationContext validationContext = new ValidationContext(SpecValidationSchema.specValidationSchema, document, i18n)
+        LanguageTraversal languageTraversal = new LanguageTraversal()
+        languageTraversal.traverse(document, new RulesVisitor(validationContext, [new DeferDirectiveOnValidOperation(validationContext, errorCollector)]))
+    }
+
+    def setup() {
+        def traversalContext = Mock(TraversalContext)
+        validationContext.getSchema() >> SpecValidationSchema.specValidationSchema
+        validationContext.getTraversalContext() >> traversalContext
+    }
+
+
+
+    def "Not allow defer on subscription operation"() {
+        given:
+        def query = """
+            subscription pets {
+                dog {
+                    ... @defer {                
+                        name 
+                    }
+                    nickname
+                }             
+            }
+        """
+
+
+        when:
+        traverse(query)
+
+        then:
+        !errorCollector.errors.isEmpty()
+        errorCollector.containsValidationError(ValidationErrorType.MisplacedDirective)
+
+    }
+
+
+    def "Allow defer(if:false) on subscription operation"() {
+        given:
+        def query = """
+            subscription pets {
+                dog {
+                    ... @defer(if:false) {                
+                        name 
+                    }
+                    nickname
+                }             
+            }
+        """
+
+
+        when:
+        traverse(query)
+
+        then:
+        errorCollector.errors.isEmpty()
+
+    }
+
+
+    def "Not allow defer on fragment when operation is subscription"() {
+        given:
+        def query = """
+            fragment doggo on PetMutationType {
+                ... { 
+                    dog {    
+                        ... @defer {
+                            id
+                        }
+                        nickname
+                    }
+                    
+                }       
+            }
+            
+            subscription doggoMutation {
+                ...{
+                    ...doggo
+                }
+           }
+            
+
+        """
+        when:
+        traverse(query)
+
+        then:
+
+        then:
+        !errorCollector.errors.isEmpty()
+        errorCollector.containsValidationError(ValidationErrorType.MisplacedDirective)
+
+    }
+
+    def "Allow defer(if:false) on fragment when operation is subscription"() {
+        given:
+        def query = """
+            fragment doggo on PetMutationType {
+                ... { 
+                    dog {    
+                        ... @defer(if:false) {
```

Comment:
can we add tests where the value of `if` is passed as a variable? And also when its value is `true`?

Reply:
added https://github.com/graphql-java/graphql-java/pull/3439/files#diff-047a30a0c2f564205e0e9c9f70108a6e50f4e15bde07165d8a754fb8e3ea5c2cR428